### PR TITLE
ARROW-17088: [R] Use `.arrow` as extension of IPC files of datasets

### PR DESCRIPTION
**Pull request by mopcup**

### CODE REVIEW

File: r/tests/testthat/test-dataset-write.R
Comment (nealrichardson) [inline r/tests/testthat/test-dataset-write.R:142]: We should have a test of providing `basename_template`, looks like we don't have one. Would you be interested in adding one here?

File: r/R/dataset-write.R
Comment (nealrichardson) [inline r/R/dataset-write.R:127]: I actually don't think you need this change due to the magic of lazy evaluation, you just need to add `format <- ifelse(as.character(format) %in% c("feather", "ipc"), "arrow", format)` after `format <- match.arg(format)`

```
> f <- function(x = c("a", "b", "c"), basename_template = paste0("part-{i}.", x)) {
+   x <- match.arg(x)
+   x <- toupper(x)
+   basename_template
+ }
> f()
[1] "part-{i}.A"
```

The benefit of this is that you still get to see the default value for `basename_template` in the signature/docs.

**Review by nealrichardson — COMMENTED** [review]
Thanks for doing this. A couple of minor suggestions but overall this looks good to me.

File: r/tests/testthat/test-dataset-write.R
Comment (mopcup) [reply r/tests/testthat/test-dataset-write.R:142]: Yes, I would.
I will try do so.

File: r/R/dataset-write.R
Comment (mopcup) [reply r/R/dataset-write.R]: I'm not familiar with lazy evaluation, and it's still difficult for me. I learned a lot. Thank you very much.

I did some updates.
please check it.

File: r/tests/testthat/test-dataset-write.R
Comment (mopcup) [reply r/tests/testthat/test-dataset-write.R:142]: I added testcases for `basename_template`.

File: r/R/dataset-write.R
Comment (eitsupi) [inline r/R/dataset-write.R:39]:
```suggestion
#' If not specified, it defaults to "part-{i}.<default extension>".
```

File: r/tests/testthat/test-dataset-write.R
Comment (eitsupi) [inline r/tests/testthat/test-dataset-write.R:474]: I think these lines should not be removed.

Comment (mopcup) [reply r/tests/testthat/test-dataset-write.R:474]: thank you.

File: r/R/dataset-write.R
Comment (nealrichardson) [inline r/R/dataset-write.R:137]: Just a stylistic thing, IMO this is more readable

```suggestion
  if (format %in% c("feather", "ipc")) {
    format <- "arrow"
  }
```

File: r/tests/testthat/test-dataset-write.R
Comment (nealrichardson) [inline r/tests/testthat/test-dataset-write.R:505]: Is this change necessary?

File: r/tests/testthat/test-dataset-write.R
Comment (nealrichardson) [inline r/tests/testthat/test-dataset-write.R:154]: We test the error mode below. How about something with a different basename_template that is valid, like "{i}.data" or something?

**Review by nealrichardson — COMMENTED** [review]
Thanks for doing this. A couple more small suggestions.

File: r/tests/testthat/test-dataset-write.R
Comment (mopcup) [reply r/tests/testthat/test-dataset-write.R:159]: Thank you very much. I have added the test case.

File: r/tests/testthat/test-dataset-write.R
Comment (mopcup) [reply r/tests/testthat/test-dataset-write.R]: sorry, my mistake.

**Review by nealrichardson — APPROVED** [review]
LGTM, thanks! Will merge when CI is green.

**Comment by mopcup** [comment]
@nealrichardson This is my first contribution to OSS. I was very anxious, but with the support of a Japanese committer, I took on the challenge. Also, I really appreciated all your comments and reviews, which motivated me.

**Comment by nealrichardson** [comment]
> @nealrichardson This is my first contribution to OSS. I was very anxious, but with the support of a Japanese committer, I took on the challenge. Also, I really appreciated all your comments and reviews, which motivated me.

Welcome to the Arrow community! We're happy you're here. I'm glad you had a good experience.

**Pull request merged**
